src/main/java/com/code_intelligence/jazzer/mutation/mutator/lang/FloatingPointMutatorFactory.java

```diff
@@ -40,31 +52,203 @@ public Optional<SerializingMutator<?>> tryCreate(AnnotatedType type, MutatorFact
+          format(
+              "[%f, %f] can not be mutated, use a constant instead: %s", minValue, maxValue, type));
+      this.minValue = minValue;
+      this.maxValue = maxValue;
+      this.allowNaN = allowNaN;
+      this.specialValues = collectSpecialValues(minValue, maxValue);
+    }
+
+    private float[] collectSpecialValues(float minValue, float maxValue) {
+      float[] specialValues = new float[] {Float.NEGATIVE_INFINITY, Float.POSITIVE_INFINITY, 0.0f,
+          -0.0f, Float.NaN, Float.MAX_VALUE, Float.MIN_VALUE, -Float.MAX_VALUE, -Float.MIN_VALUE,
+          this.minValue, this.maxValue};
+      ArrayList<Float> specialValuesList = new ArrayList<>();
+      for (float specialValue : specialValues) {
+        if ((allowNaN && Float.isNaN(specialValue))
+            || (specialValue >= minValue && specialValue <= maxValue)) {
+          specialValuesList.add(specialValue);
+        }
+      }
+      float[] specialValuesArray = new float[specialValuesList.size()];
+      for (int i = 0; i < specialValuesList.size(); i++) {
+        specialValuesArray[i] = specialValuesList.get(i);
+      }
+      return specialValuesArray;
+    }
 
     public Float mutateWithLibFuzzer(Float value) {
       return LibFuzzerMutator.mutateDefault(value, this, 0);
     }
 
-    // TODO: randomly init to special values like NaN, -0, +0, +Inf, -Inf
     @Override
     public Float init(PseudoRandom prng) {
-      return prng.closedRange(-Float.MAX_VALUE, Float.MAX_VALUE);
+      if (prng.choice()) {
+        return specialValues[prng.closedRange(0, specialValues.length - 1)];
+      } else {
+        return prng.closedRange(minValue, maxValue);
+      }
     }
 
-    // TODO: consider current value when mutating
     @Override
     public Float mutate(Float value, PseudoRandom prng) {
-      return prng.closedRange(-Float.MAX_VALUE, Float.MAX_VALUE);
+      float result;
+      // small chance to return a special value
+      if (prng.closedRange(0, INV_CHANCE_SPECIAL_VALUE) == 0) {
+        result = specialValues[prng.closedRange(0, specialValues.length - 1)];
+      } else {
+        switch (prng.closedRange(0, 5)) {
+          case 0:
+            result = mutateWithBitFlip(value, prng);
+            break;
+          case 1:
+            result = mutateExponent(value, prng);
+            break;
+          case 2:
+            result = mutateMantissa(value, prng);
+            break;
+          case 3:
+            result = (float) mutateWithMathematicalFn(value, prng);
+            break;
+          case 4:
+            result = mutateWithLibFuzzer(value);
+            break;
+          default: // random in range cannot exceed the given bounds (and cannot be NaN)
+            return (prng.closedRange(minValue, maxValue));
+        }
+      }
+      result = forceInRange(result, minValue, maxValue, allowNaN);
+      // If value repeats, or is a not allowed NaN, mutate again
+      if (result == value || (Float.isNaN(result) && Float.isNaN(value))
```

Comment:
Not entirely sure about edge cases, but couldn't this line be replaced with `Double.compare`?

Reply:
It works. It also revealed some bugs, since 0.0 == -0.0 is true; and  -1 when we use Double.compare. The StressTest had to be adjusted with another workaround.